Navigation CMS

MatthewKennedy opened the pull request:
**Drag & Drop Sortable Menus**

MatthewKennedy committed:
Add Spree::Menu and Spree::MenuItem

MatthewKennedy committed:
Add multiple stores support to Spree::Menu

MatthewKennedy committed:
Add link switcher, and use Awesome Nested Set

MatthewKennedy committed:
Update menu item to allow or disallow nesting.

MatthewKennedy committed:
Add drag and drop

MatthewKennedy committed:
Improve JavaScript and SCSS for SortableJS menu boilerplate.

MatthewKennedy committed:
Use generated menu_items

MatthewKennedy committed:
Update menu leveling.

MatthewKennedy committed:
Improve nesting

MatthewKennedy committed:
Update nested set.

MatthewKennedy committed:
Tidy up and add set js vars for move.

MatthewKennedy committed:
Clean up code

MatthewKennedy committed:
Fix MenuItems Tree.

MatthewKennedy committed:
Fix delete menu item jQuery fadeOut.

MatthewKennedy committed:
Add images to the menu items index.

MatthewKennedy committed:
Add menu sample data

MatthewKennedy committed:
Add ability to select Product or Taxon and generate destination output.

MatthewKennedy committed:
Update sample and add method for child index.

damianlegawiec commented on backend/app/controllers/spree/admin/menus_controller.rb:
```diff
@@ -0,0 +1,26 @@
+
+      private
+
+      def load_data
```
@MatthewKennedy I think we can have `load_stores` before_action for all controllers in the admin, but maybe this should be changed to:

```ruby
Spree::Store.accessible_by(current_ability, :show)
```

damianlegawiec commented on backend/app/controllers/spree/admin/menus_controller.rb:
```diff
@@ -0,0 +1,26 @@
+      end
+
+      def load_menu_items_ordered
+        @menu_items_in_order = @menu.menu_items.order('lft ASC')
```
```suggestion
        @menu_items_in_order = @menu.menu_items.order(:lft)
```

damianlegawiec commented on backend/app/views/spree/admin/menu_items/_form.html.erb:
```diff
@@ -0,0 +1,97 @@
+        <div class="card-body">
+          <%= f.field_container :item_type, class: ['form-group'] do %>
+            <%= f.label :item_type, Spree.t('admin.navigation.item_type') %>
+            <%= f.select(:item_type, Spree::MenuItem::ITEM_TYPE, {include_blank: false}, { class: 'select2' }) %>
```
```suggestion
            <%= f.select(:item_type, Spree::MenuItem::ITEM_TYPE, {include_blank: false}, { class: 'select2' }) %>
```

let's move that `Spree::MenuItem::ITEM_TYPE` from the views to controllers please

damianlegawiec commented on core/db/migrate/20210407200948_create_spree_menus.rb:
```diff
@@ -0,0 +1,9 @@
+class CreateSpreeMenus < ActiveRecord::Migration[6.0]
+  def change
+    create_table :spree_menus do |t|
+      t.column :name, :string
```
@MatthewKennedy let's add an index on name

damianlegawiec commented on core/db/migrate/20210408092939_create_spree_menu_items.rb:
```diff
@@ -0,0 +1,24 @@
+      t.column :item_type, :string
+      t.column :linked_resource_type, :string, default: 'URL'
+      t.column :linked_resource_id, :integer
+      t.column :parent_id, :integer
```
index here

damianlegawiec commented on core/db/migrate/20210408092939_create_spree_menu_items.rb:
```diff
@@ -0,0 +1,24 @@
+      t.column :linked_resource_type, :string, default: 'URL'
+      t.column :linked_resource_id, :integer
+      t.column :parent_id, :integer
+      t.column :lft, :integer
```
and here

damianlegawiec commented on core/db/migrate/20210408092939_create_spree_menu_items.rb:
```diff
@@ -0,0 +1,24 @@
+      t.column :linked_resource_id, :integer
+      t.column :parent_id, :integer
+      t.column :lft, :integer
+      t.column :rgt, :integer
```
and here

damianlegawiec commented on core/db/migrate/20210408092939_create_spree_menu_items.rb:
```diff
@@ -0,0 +1,24 @@
+      t.column :url, :string
+      t.column :new_window, :boolean, default: false
+      t.column :item_type, :string
+      t.column :linked_resource_type, :string, default: 'URL'
+      t.column :linked_resource_id, :integer
```
index on both fields please

MatthewKennedy replied:
I'll `add_index :spree_menu_items, [:linked_resource_type, :linked_resource_id]` to optimise searching for a Taxon and  it's ID or a Product and it's ID, or Page and it's ID

<img width="1175" alt="Screenshot 2021-04-15 at 11 18 29" src="https://user-images.githubusercontent.com/1240766/114854114-64890780-9ddc-11eb-914f-ef4bfba29224.png">
<img width="1193" alt="Screenshot 2021-04-15 at 11 18 42" src="https://user-images.githubusercontent.com/1240766/114854127-6783f800-9ddc-11eb-9e70-f6b2c90bbd5e.png">

damianlegawiec commented on core/db/migrate/20210408092939_create_spree_menu_items.rb:
```diff
@@ -0,0 +1,24 @@
+      t.column :subtitle, :string
+      t.column :url, :string
+      t.column :new_window, :boolean, default: false
+      t.column :item_type, :string
```
index :)

damianlegawiec commented on sample/db/samples/menu_items.rb:
```diff
@@ -0,0 +1,240 @@
+#################################
+Spree::MenuItem.where(
+  name: 'Skirts',
+  linked_resource_type: 'URL',
```
@MatthewKennedy can we link to taxon please?

MatthewKennedy replied:
Absolutely, I will give that a good going over today and get that sample data all correct.

MatthewKennedy committed:
Tidy code and add more indexes.

damianlegawiec commented on core/app/models/spree/menu_item.rb:
```diff
@@ -0,0 +1,67 @@
+
+    def destination
+      if DYNAMIC_RESOURCE_TYPE.include? linked_resource_type
+        return if linked_resource_id.nil?
+
```
```suggestion
        return if linked_resource.nil?

```

damianlegawiec commented on core/app/models/spree/menu_item.rb:
```diff
@@ -0,0 +1,67 @@
+
+        case linked_resource_type
+        when 'Taxon'
+          permalink = Spree::Taxon.find(linked_resource_id).permalink || 'summer'
+          Spree::Core::Engine.routes.url_helpers.nested_taxons_path(permalink)
```
```suggestion
          Spree::Core::Engine.routes.url_helpers.nested_taxons_path(linked_resource.permalink)
```

damianlegawiec commented on core/app/models/spree/menu_item.rb:
```diff
@@ -0,0 +1,67 @@
+          permalink = Spree::Taxon.find(linked_resource_id).permalink || 'summer'
+          Spree::Core::Engine.routes.url_helpers.nested_taxons_path(permalink)
+        when 'Product'
+          product = Spree::Product.find(linked_resource_id)
+          Spree::Core::Engine.routes.url_helpers.product_path(product)
```
```suggestion
          Spree::Core::Engine.routes.url_helpers.product_path(linked_resource)
```

MatthewKennedy committed:
Set sample data to use taxon links.

MatthewKennedy committed:
Set Main Menu and Footer Menu links in sample data.

MatthewKennedy committed:
Use polymorphic associations for linked_resources. Make sure Menu name is unique for indexing.

MatthewKennedy committed:
Check for linked_resource.nil?

MatthewKennedy committed:
Merge remote-tracking branch 'origin/master' into navigation

MatthewKennedy committed:
Add some feature tests and some missing translations.

MatthewKennedy committed:
Clean up tests wording.

MatthewKennedy committed:
Update tests, small visual change.

MatthewKennedy committed:
Add "used in" section to menus table to show what stores each menu is used in.

MatthewKennedy committed:
Add model test for Spree::Menu

MatthewKennedy committed:
Start on menu item model spec.

MatthewKennedy committed:
Hook up drag and drop re-positioning.

MatthewKennedy committed:
Remove added line in base helper for core specs

MatthewKennedy committed:
Improve JS.

MatthewKennedy committed:
remove log output line.

MatthewKennedy committed:
Add flash error message if the menu_item move is not saved.

MatthewKennedy committed:
Implement the front end desktop Main Menu to use the new navigation menus system.

MatthewKennedy committed:
Resolve an issue with menus when the Spree::Frontend is not defined.

MatthewKennedy committed:
Update js

MatthewKennedy committed:
small tweaks

MatthewKennedy committed:
revert change to edit and delete helper

MatthewKennedy committed:
Refactor of assets to all be under the name space of menu / menus.

MatthewKennedy committed:
Add :unique_code to Spree::Menu and :code to Spree::MenuItem allowing for sturdy targeting, even if the :name is translated, or simply changed.

MatthewKennedy committed:
Use Spree::Menu for Footer links.

MatthewKennedy committed:
Use root not item in main menu.

MatthewKennedy committed:
Convert mobile navigation to use Spree::Menu

MatthewKennedy committed:
Sett a fallback for a blank URL

MatthewKennedy committed:
Add unique code to tests.

MatthewKennedy committed:
Fix tests.

MatthewKennedy committed:
Simplify drag and drop menu builder.

MatthewKennedy committed:
Set locale spec to look for menu items.

MatthewKennedy committed:
Add more info to the menu item bars so user can distinguish between them if the name is the same.

MatthewKennedy committed:
Adjust main_nav_bar to allow creation of multiple categories.

MatthewKennedy commented:
Adds ability to create menus with 
<img width="1280" alt="Screenshot 2021-04-21 at 10 24 23" src="https://user-images.githubusercontent.com/1240766/115531288-82e37d00-a28c-11eb-94df-2f1232dee503.png">

MatthewKennedy committed:
Use path

MatthewKennedy committed:
use &:save!

MatthewKennedy committed:
Update code

MatthewKennedy committed:
Remove unused argument.

MatthewKennedy committed:
Update to use destination

MatthewKennedy committed:
Update view.

MatthewKennedy committed:
Change default path to t

MatthewKennedy committed:
Fix test and localize links.

MatthewKennedy committed:
Merge remote-tracking branch 'origin/master' into navigation

MatthewKennedy committed:
Add menu API.

MatthewKennedy committed:
update mobile view.

MatthewKennedy committed:
Move menu items to a dedicated root item.

MatthewKennedy committed:
Fix footer menu and mobile menu.

MatthewKennedy committed:
Fix test by using dedicated root item, improve code.

MatthewKennedy committed:
Remove moving to root form API.

MatthewKennedy committed:
Small tweaks.

MatthewKennedy committed:
Add more tests.

MatthewKennedy committed:
Add tests, fix some CC issues.

MatthewKennedy committed:
Add test for helper method.

MatthewKennedy committed:
Improve using menu items in frontend.

MatthewKennedy committed:
Add more tests.

MatthewKennedy committed:
Refactor menu item image

MatthewKennedy committed:
Merge branch 'master' into navigation

MatthewKennedy committed:
Finish storefront API V2 for menu and ?include=menu_items.

MatthewKennedy committed:
Remove old .yml navigation install files.

MatthewKennedy committed:
Merge branch 'master' into navigation

damianlegawiec commented on frontend/app/helpers/spree/navigation_helper.rb:
```diff
@@ -2,17 +2,9 @@
 
 module Spree
   module NavigationHelper
-    def spree_navigation_data
```
@MatthewKennedy can we not remove this method as well as the initializer? this will ease the migration as developers will still be able to use this method with old templates and migrate to new nav cms gradually. We definitely should deprecate this method and remove it in Spree 5.0

damianlegawiec commented on frontend/lib/generators/spree/frontend/install/install_generator.rb:
```diff
@@ -16,9 +16,6 @@ def self.source_paths
         end
 
         def install
-          # main navigation configuration
-          template 'config/initializers/spree_storefront.rb', 'config/initializers/spree_storefront.rb'
-          template 'config/spree_storefront.yml', 'config/spree_storefront.yml'
```
This is fine :) For new installations we want to dispose of those files, we can also remove all of the https://github.com/spree/spree/tree/master/frontend/app/assets/images/meganav

damianlegawiec commented on frontend/spec/helpers/navigation_helper_spec.rb:
```diff
@@ -5,28 +5,6 @@ module Spree
     include Spree::CurrencyHelper
     include Spree::LocaleHelper
 
-    describe '#spree_navigation_data' do
```
@MatthewKennedy let's keep the test as well :)

MatthewKennedy replied:
Sure, and I'll add some for the new methods this afternoon.

damianlegawiec commented on backend/app/controllers/spree/admin/base_controller.rb:
```diff
@@ -8,6 +8,7 @@ class BaseController < Spree::BaseController
 
       before_action :authorize_admin
       before_action :generate_admin_api_key
+      before_action :load_data
```
```suggestion
      before_action :load_stores
```

damianlegawiec commented on backend/app/controllers/spree/admin/base_controller.rb:
```diff
@@ -49,6 +50,10 @@ def config_locale
         Spree::Backend::Config[:locale]
       end
 
+      def load_data
```
```suggestion
      def load_stores
```

damianlegawiec commented on backend/app/controllers/spree/admin/base_controller.rb:
```diff
@@ -49,6 +50,10 @@ def config_locale
       end
 
+      def load_data
+        @stores = Spree::Store.accessible_by(current_ability, :show)
```
```suggestion
        @stores = Spree::Store.accessible_by(current_ability, :show).order(default: :desc)
```

damianlegawiec commented on backend/app/controllers/spree/admin/menu_items_controller.rb:
```diff
@@ -0,0 +1,25 @@
+        @menu_item_types = Spree::MenuItem::ITEM_TYPE
+        @linked_resource_types = Spree::MenuItem::LINKED_RESOURCE_TYPE
+
+        @taxons = Spree::Taxon.all
+        @products = Spree::Product.all
```
@MatthewKennedy we need to rather use Platform API endpoints to fetch Products and Taxons, loading so many records into the memory will slow things down

damianlegawiec commented on core/app/models/spree/menu_item.rb:
```diff
@@ -0,0 +1,84 @@
+    LINKED_RESOURCE_TYPE.push(*DYNAMIC_RESOURCE_TYPE)
+
+    validates :name, presence: true
+    validates :menu_id, presence: true, numericality: { only_integer: true }
```
```suggestion
    validates :menu, presence: true
```

let's rely on association presence validation, also if someone would like to use UUIDs this would fail

damianlegawiec commented on core/app/models/spree/menu_item.rb:
```diff
@@ -0,0 +1,84 @@
+        self.destination = if frontend_available?
+                             Spree::Core::Engine.routes.url_helpers.nested_taxons_path(linked_resource.permalink)
+                           else
+                             "/#{Spree::Config[:api_taxon_path]}/#{linked_resource.permalink}"
```
```suggestion
                             "/#{Spree::Config[:storefront_taxon_path]}/#{linked_resource.permalink}"
```

damianlegawiec commented on core/app/models/spree/menu_item.rb:
```diff
@@ -0,0 +1,84 @@
+        self.destination = if frontend_available?
+                             Spree::Core::Engine.routes.url_helpers.product_path(linked_resource)
+                           else
+                             "/#{Spree::Config[:api_product_path]}/#{linked_resource.slug}"
```
```suggestion
                             "/#{Spree::Config[:storefront_product_path]}/#{linked_resource.slug}"
```

damianlegawiec commented on core/app/models/spree/menu_item.rb:
```diff
@@ -0,0 +1,84 @@
+    end
+
+    def paremeterize_code
+      if code
+        self.code = code.parameterize
+      end
```
```suggestion
      return if code.blank?
      
      self.code = code.parameterize
```

damianlegawiec commented on core/app/models/spree/menu_item_image.rb:
```diff
@@ -0,0 +1,9 @@
+  class MenuItemImage < Spree::Asset
+    has_one_attached :attachment
+
+    validates :attachment, attached: true, content_type: %i[png jpg jpeg gif svg]
```
let's move `%i[png jpg jpeg gif svg]` into constant

damianlegawiec commented on core/app/models/spree/product.rb:
```diff
@@ -94,6 +96,8 @@ class Product < Spree::Base
     after_save :reset_memoized_data
     after_commit :reset_memoized_data
 
+    around_update :sync_menu_item_paths
```
@MatthewKennedy can we move this to `after_save/after_commit`?

damianlegawiec commented on core/app/models/spree/taxon.rb:
```diff
@@ -33,6 +35,8 @@ class Taxon < Spree::Base
     after_save :touch_ancestors_and_taxonomy
     after_touch :touch_ancestors_and_taxonomy
 
+    around_update :sync_menu_item_paths
```
same here

MatthewKennedy committed:
Improve code as per the code reviews.

MatthewKennedy committed:
Use Platform API for fetching and populating select options.

MatthewKennedy committed:
Update JavaScript to handle errors.

MatthewKennedy committed:
Remove unused argument. Use Platform API.

damianlegawiec commented on api/app/controllers/spree/api/v2/platform/products_controller.rb:
```diff
@@ -0,0 +1,55 @@
+          end
+
+          def collection
+            @collection ||= collection_finder.new(scope: scope, params: params, current_currency: current_currency).execute
```
@MatthewKennedy let's not use custom finders in Platform API at all and rely on ransack

damianlegawiec commented on api/app/controllers/spree/api/v2/platform/products_controller.rb:
```diff
@@ -0,0 +1,55 @@
+          private
+
+          def sorted_collection
+            collection_sorter.new(collection, current_currency, params, allowed_sort_attributes).call
```
@MatthewKennedy also we don't need custom sorters, ransack handles it well

damianlegawiec commented on api/app/controllers/spree/api/v2/platform/taxons_controller.rb:
```diff
@@ -0,0 +1,47 @@
+          end
+
+          def collection_finder
+            Spree::Api::Dependencies.storefront_taxon_finder.constantize
```
@MatthewKennedy same here :)

damianlegawiec commented on api/app/serializers/spree/api/v2/platform/product_serializer.rb:
```diff
@@ -0,0 +1,69 @@
+
+          set_type :product
+
+          attributes :name, :description, :available_on, :slug, :meta_description, :meta_keywords, :updated_at
+
+          attribute :purchasable do |product|
+            product.purchasable?
+          end
+
+          attribute :in_stock do |product|
+            product.in_stock?
+          end
+
+          attribute :backorderable do |product|
+            product.backorderable?
+          end
+
+          attribute :available do |product|
+            product.available?
+          end
+
+          attribute :currency do |_product, params|
+            params[:currency]
+          end
```
@MatthewKennedy for Platform API you only need to add `include ::Spree::Api::V2::ResourceSerializerConcern`

MatthewKennedy replied:
I'll go over this again, and get it in good shape.

MatthewKennedy committed:
Clean up newly added controllers and serializes.

MatthewKennedy commented on api/app/serializers/spree/api/v2/platform/taxon_serializer.rb:
```diff
@@ -0,0 +1,38 @@
+  module Api
+    module V2
+      module Platform
+        class TaxonSerializer < BaseSerializer
```
I could not use `include ::Spree::Api::V2::ResourceSerializerConcern` here, is this because a taxon is self nesting?

MatthewKennedy committed:
Only sync a menu item destination path if it is changed.

MatthewKennedy committed:
Tidy up JavaScript + allow users to clear selected taxon or product.

MatthewKennedy committed:
Add test for clearing selection

MatthewKennedy committed:
Merge branch 'master' into navigation

damianlegawiec commented on core/db/migrate/20210407200948_create_spree_menus.rb:
```diff
@@ -0,0 +1,13 @@
+    end
+
+    add_index :spree_menus, :name
+    add_index :spree_menus, :unique_code, unique: true
```
```suggestion
    add_index :spree_menus, :unique_code
```

let's have only rails unique validation here

damianlegawiec commented on api/app/serializers/spree/v2/storefront/menu_serializer.rb:
```diff
@@ -0,0 +1,13 @@
+      class MenuSerializer < BaseSerializer
+        set_type :menu
+
+        attributes :name, :unique_code, :store_ids
```
I think we should add `have_many: :menus` to `Storefront::StoreSerializer`

MatthewKennedy replied:
Will sort this.

damianlegawiec committed:
Full path when editing Menu Item

damianlegawiec committed:
Added :reposition to global WRITE actions of Platform API

This is a common action actoss Spree models such as Taxons, Products inside Taxons, etc so it can be moved to ResourceController

MatthewKennedy commented on core/app/models/spree/menu.rb:
```diff
@@ -0,0 +1,27 @@
+    after_create :set_root
+
+    validates :name, presence: true
+    validates :unique_code, presence: true, uniqueness: true
```
No problem.

damianlegawiec commented:
@MatthewKennedy 

Two major changes needed: 

<img width="1159" alt="Screenshot 2021-05-02 at 21 01 56" src="https://user-images.githubusercontent.com/55154/116824287-ad391280-ab89-11eb-94c1-53baa3d197e6.png">

This need to be an autocomplete field, we cannot load every product via XHR, we need to fetch only the products that match the query - you can check (and rewrite) https://github.com/spree/spree/blob/master/backend/app/assets/javascripts/spree/backend/product_picker.js

The same goes for Taxon dropdown

MatthewKennedy commented:
@damianlegawiec I will change those to use autocomplete.

MatthewKennedy committed:
Update api

MatthewKennedy committed:
Add custom validation for unique_code within scope of stores.

MatthewKennedy commented on api/app/controllers/spree/api/v2/platform/menu_items_controller.rb:
```diff
@@ -0,0 +1,35 @@
+    module V2
+      module Platform
+        class MenuItemsController < ResourceController
+          before_action -> { doorkeeper_authorize! :write, :admin }, only: WRITE_ACTIONS << :reposition
```
@damianlegawiec I had to put this back in, for some reason moving `:reposition` to the `WRITE_ACTIONS` constant in the `ResourceController ` stopped the reposition action working, I'll look into that more and see if I can find out exactly why.

MatthewKennedy committed:
Add tests for new scoped unique validation.

MatthewKennedy committed:
Remove meganav assets generator.

damianlegawiec commented on core/app/models/spree/menu.rb:
```diff
@@ -0,0 +1,45 @@
+    before_validation :paremeterize_unique_code
+    after_create :set_root
+
+    validate :uniqueness_of_code_within_stores
```
```suggestion
    validate : unique_code, uniqueness: { scope: :store_id }
```

that should do the trick

MatthewKennedy replied:
That was my first try: `validates :unique_code, uniqueness: { scope: :store_id }`

But because it is a join table many to many with a through association, it's looking for the `store_id` on the `spree_menus` table, then that exists on the `spree_menus_stores` table.

<img width="943" alt="Screenshot 2021-05-04 at 07 59 26" src="https://user-images.githubusercontent.com/1240766/116969900-f3d45d00-acae-11eb-8760-73e420c43bc5.png">

Do you want me to convert this to a has_one relationship?

damianlegawiec replied:
@MatthewKennedy yes, converting to `belongs_to` is the way to go, this will simplify things in multi-store/cloud scenario a lot

MatthewKennedy replied:
@damianlegawiec No problem, I'll sort that.

MatthewKennedy committed:
Add Select2 Autocomplete for Products & Taxon.

MatthewKennedy committed:
Convert to belons_to relationship.

damianlegawiec commented on api/app/serializers/spree/api/v2/platform/menu_item_serializer.rb:
```diff
@@ -5,6 +5,20 @@ module Platform
         class MenuItemSerializer < BaseSerializer
           include ::Spree::Api::V2::ResourceSerializerConcern
 
+          attributes :item_type, :code, :name, :subtitle, :destination, :new_window, :lft, :rgt, :depth
```
@MatthewKennedy we can skip this line as `include ::Spree::Api::V2::ResourceSerializerConcern` adds those attributes automatically :)

damianlegawiec commented on api/app/serializers/spree/api/v2/platform/menu_serializer.rb:
```diff
@@ -5,6 +5,9 @@ module Platform
         class MenuSerializer < BaseSerializer
           include ::Spree::Api::V2::ResourceSerializerConcern
 
+          attributes :name, :unique_code
```
@MatthewKennedy same here

damianlegawiec commented on api/app/serializers/spree/api/v2/platform/store_serializer.rb:
```diff
@@ -0,0 +1,18 @@
+        class StoreSerializer < BaseSerializer
+          include ::Spree::Api::V2::ResourceSerializerConcern
+
+          attributes :name, :url, :meta_description, :meta_keywords, :seo_title, :default_currency, :default, :supported_currencies, :facebook,
+                     :twitter, :instagram, :default_locale, :customer_support_email, :default_country_id, :description,
+                     :address, :contact_phone, :supported_locales
```
@MatthewKennedy and here :)

damianlegawiec commented on backend/app/views/spree/admin/menus/index.html.erb:
```diff
@@ -24,8 +24,8 @@
           <td>
-            <% menu.store_ids.each do |store| %>
-              <span class="badge badge-secondary"><%= @stores.find(store).unique_name %></span>
+            <% if menu.store_id %>
```
we need to add `store` presence validation, there should not be a `menu` record without store assigned

damianlegawiec commented on backend/app/views/spree/admin/menus/index.html.erb:
```diff
@@ -24,8 +24,8 @@
-            <% menu.store_ids.each do |store| %>
-              <span class="badge badge-secondary"><%= @stores.find(store).unique_name %></span>
+            <% if menu.store_id %>
+              <span class="badge badge-secondary"><%= @stores.find(menu.store_id).unique_name %></span>
```
`menu.store.unique_name` should be better for the eyes :)

MatthewKennedy replied:
I will sort all of the above, nothing too tricky to fix there.

MatthewKennedy committed:
Fix code reviews, Fix tests.

MatthewKennedy committed:
Fix broken tests, improve menus index.

MatthewKennedy commented:
Current State of the indexes For Reference:
**spree_menu Table**
```Ruby
    t.index ["name"], name: "index_spree_menus_on_name"
    t.index ["store_id"], name: "index_spree_menus_on_store_id"
    t.index ["unique_code"], name: "index_spree_menus_on_unique_code"
```

**spree_menu_items**
```Ruby
    t.index ["code"], name: "index_spree_menu_items_on_code"
    t.index ["depth"], name: "index_spree_menu_items_on_depth"
    t.index ["item_type"], name: "index_spree_menu_items_on_item_type"
    t.index ["lft"], name: "index_spree_menu_items_on_lft"
    t.index ["linked_resource_type", "linked_resource_id"], name: "index_spree_menu_items_on_linked_resource"
    t.index ["menu_id"], name: "index_spree_menu_items_on_menu_id"
    t.index ["parent_id"], name: "index_spree_menu_items_on_parent_id"
    t.index ["rgt"], name: "index_spree_menu_items_on_rgt"
```

MatthewKennedy committed:
Use API to V2 for new autocomplete.

MatthewKennedy committed:
Update to use platform API on admin.

MatthewKennedy committed:
Fix using taxon API URL in both scenarios.

MatthewKennedy committed:
Fix filtering.

damianlegawiec commented on core/app/models/spree/menu_item.rb:
```diff
@@ -0,0 +1,84 @@
+    validates :linked_resource_type, inclusion: { in: LINKED_RESOURCE_TYPE }
+    validate :check_for_root, on: :create
+
+    has_one :menu_item_image, as: :viewable, dependent: :destroy, class_name: 'Spree::MenuItemImage'
```
@MatthewKennedy can we skip the additional model here and just have `has_one_attached` ?

MatthewKennedy replied:
@damianlegawiec I'm happy to change this back to `has_one_attached`, it is the simplest solution and the one I went with first, just a consideration before I go ahead.

Retrieving the image through the json api serializer, I only see this done using the external model way.

I can change this to use the existing `Spree::Image` model rather than `Spree::MenuItem::Image` model, and still get the image through the serializer for the API.

What do yo think is best?

MatthewKennedy replied:
There is also a bug with awesome nested set and active storage

If you use `acts_as_nested_set` and `has_one_attached` in the same model the attachment is lost if you attach one to a newly created item.

The bug is not present when use use active storage via a separate model, in the current way.

damianlegawiec commented on core/app/helpers/spree/base_helper.rb:
```diff
@@ -25,6 +25,24 @@ def display_price(product_or_variant)
         to_html
     end
 
+    def localized_item_link(item, locale)
```
@MatthewKennedy I think this can be moved to frontends `NavigationHelper`, also let's prefix it with `spree_`

damianlegawiec commented on core/app/helpers/spree/base_helper.rb:
```diff
@@ -25,6 +25,24 @@ def display_price(product_or_variant)
+    def localized_item_link(item, locale)
+      return if item.destination.nil?
+
+      output_locale = if locale == current_store.default_locale
```
@MatthewKennedy you can use `locale_param` for this - https://github.com/spree/spree/blob/c47e591882d79e766847c627773d29186aa2224b/core/lib/spree/core/controller_helpers/locale.rb#L50

MatthewKennedy committed:
Update code reviews.

damianlegawiec commented on frontend/app/views/spree/shared/_main_nav_bar.html.erb:
```diff
@@ -1,55 +1,71 @@
-<% if spree_navigation_data.any? %>
+<% if spree_top_level_items_for_menu('spree-all-main').present? %>
```
@MatthewKennedy this cannot be hardcoded - maybe let's also add `locale` column to `Menu` model and find that menus dynamically:

```
current_store.menus.for_locale(I18n.locale) || current_store.menus.first
```

MatthewKennedy replied:
Sure, no problem

damianlegawiec replied:
actually, it should be 

```ruby
current_store.menus.for_locale(I18n.locale).for_footer || current_store.menus.for_footer.first
```

we should have a helper method

```
def spree_menu(kind = `header`)
  scope = current_store.menus.for_kind(kind)

  scope.for_locale(I18n.locale) || scope.first
end
```

MatthewKennedy replied:
Noted!

MatthewKennedy committed:
Code reviews

MatthewKennedy committed:
Multiple language menus.

damianlegawiec commented on frontend/app/views/spree/shared/_footer.html.erb:
```diff
@@ -49,8 +49,8 @@
 
-        <% if spree_top_level_items_for_menu('spree-all-footer').present? %>
-          <% spree_top_level_items_for_menu('spree-all-footer').each do |parent| %>
+        <% if available_menus.for_footer(I18n.locale).present? %>
```
```suggestion
        <% if spree_footer_menu.present? %>
```

let's keep as less ruby code from the templates as possible, let's move this into the navigation helper

damianlegawiec commented on core/db/migrate/20210407200948_create_spree_menus.rb:
```diff
@@ -2,14 +2,16 @@ class CreateSpreeMenus < ActiveRecord::Migration[6.0]
     add_index :spree_menus, :name
-    add_index :spree_menus, :unique_code
+    add_index :spree_menus, :location
+    add_index :spree_menus, :locale
```
```suggestion
    add_index :spree_menus, :locale
    add_index :spree_menus, [:store_id, :location, :locale], unique: true
```

damianlegawiec commented on frontend/app/views/spree/shared/_footer.html.erb:
```diff
@@ -49,8 +49,8 @@
-        <% if spree_top_level_items_for_menu('spree-all-footer').present? %>
-          <% spree_top_level_items_for_menu('spree-all-footer').each do |parent| %>
+        <% if available_menus.for_footer(I18n.locale).present? %>
+          <% available_menus.for_header(I18n.locale).children.each do |parent| %>
```
```suggestion
          <% spree_footer_menu.children.each do |parent| %>
```

MatthewKennedy committed:
Sync menu name to root item name, allowing user to get menu name with spree_menu('footer').name

MatthewKennedy committed:
Clean up view files for main nav.

MatthewKennedy committed:
Fix tests and add API test for menu_item#reposition

MatthewKennedy committed:
Improve select options in view.

MatthewKennedy committed:
Fix CC issue and Note TODO

MatthewKennedy commented on api/app/controllers/spree/api/v2/platform/menu_items_controller.rb:
```diff
@@ -0,0 +1,36 @@
+
+          def reposition
+            # TODO: In the tests I can't seem to get past this.
+            # spree_authorize! :update, @moved_item
```
@damianlegawiec Thanks, I'll try that shortly.

MatthewKennedy committed:
Use view helpers to format arrays for select.

MatthewKennedy committed:
Add the ability to filter and search menus in admin UI.

MatthewKennedy committed:
Add formatter to output stores unique name in select.

damianlegawiec commented on backend/app/helpers/spree/admin/base_helper.rb:
```diff
@@ -218,6 +218,16 @@ def admin_logout_link
         end
       end
 
+      def stores_formatted_for_select
```
@MatthewKennedy let's add this to `Spree::Admin::StoresHelper` - there will be more stores-related methods

damianlegawiec commented on backend/app/helpers/spree/admin/base_helper.rb:
```diff
@@ -218,6 +218,16 @@ def admin_logout_link
+        formatted_stores = []
+
+        @stores.each do |store|
+          formatted_stores << ["#{store.name} (#{store.code})", store.id]
```
@MatthewKennedy you can use `store.unique_name` for this

damianlegawiec commented on backend/app/helpers/spree/admin/base_helper.rb:
```diff
@@ -218,6 +218,16 @@ def admin_logout_link
+      def stores_formatted_for_select
+        formatted_stores = []
+
+        @stores.each do |store|
```
also you can switch from `each` to `map`

damianlegawiec commented on backend/app/helpers/spree/admin/base_helper.rb:
```diff
@@ -218,6 +218,16 @@ def admin_logout_link
         end
       end
 
+      def stores_formatted_for_select
```
```suggestion
      def stores_dropdown_values
```

MatthewKennedy replied:
Going to submit this as a separate commit to test the random failing test.

MatthewKennedy committed:
Fix code reviews.

MatthewKennedy committed:
Add language name for failing test and rename the select options helpers to use naming convention: _dropdown_values

damianlegawiec committed:
Added LocaleHelper#locale_full_name

damianlegawiec reviewed (approved):
🚀 🚀 🚀
